packages/runtime/container-runtime/src/blobManager.ts

```diff
@@ -391,13 +397,21 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
 		);
 	}
 
-	private getBlobHandle(id: string): IFluidHandle<ArrayBufferLike> {
+	private getBlobHandle(id: string, isOnline: boolean): BlobHandle {
 		assert(
 			this.redirectTable.has(id) || this.pendingBlobs.has(id),
 			0x384 /* requesting handle for unknown blob */,
 		);
-		return new BlobHandle(`${BlobManager.basePath}/${id}`, this.routeContext, async () =>
-			this.getBlob(id),
+
+		return new BlobHandle(
+			`${BlobManager.basePath}/${id}`,
+			this.routeContext,
+			async () => this.getBlob(id),
+			() => {
+				if (isOnline) {
```

Comment:
i don't think isOnline is right here, as that state can change between handle creation, and attach. you need to evaluate the whole pending state object

Reply:
you also can't just remove the code below, as attach can come before upload, so in that case you need to delete.

I would probably write a new function that is called here, and in processBlobAttachOp that takes the id, gets it from the pending, and detemines if the blob is no longer pending and can be deleted

Reply:
the key is that there are three states, uploaded, op round tripped, and attached, all three need to be true to be not pending anymore.